### PR TITLE
Optimize EDS stream allocation

### LINKED ISSUE #2943: Optimize EDS stream allocation
Today, EDS attempts to open one stream per cluster to the management server when not being multiplexed over ADS. EDS streams support subscribing to more than one cluster, so this should not be necessary if the `ApiConfigSource` points to the same management server.

This is actually very noticeable when we have a large number of clusters. It shows up as stream resource exhaustion in H2 (https://github.com/envoyproxy/envoy/issues/2941) as well as in internal performance benchmarks of configuration scalability.

The fix seems pretty easy, just need to do some level of aggregation similar to what ADS does.

### PR DESCRIPTION
**Pull request by nezdolik**
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Commit Message: Optimize streams allocation for Eds Grpc
Additional Description: With this change Eds Grpc subscriptions towards the same management server reuse a shared grpc connection, which greatly reduces the number of concurrent H2 streams towards a single management server. The work is built on top of [5026](https://github.com/envoyproxy/envoy/pull/5026), with bugs fixed and an addition of integration test for Eds over Grpc. I could not port 3 existing Eds integration tests for warming clusters, as they were blocked by [non graceful shutdown of GoogleAsyncClient](https://github.com/envoyproxy/envoy/issues/15072). 

The change has also been tested in production Spotify environment, we use Eds Grpc. Envoy successfully resolved cluster load assignments via optimized Eds.

Risk Level: High
Testing: Unit tests + integration tests
Docs Changes: TBD
Release Notes: TBD
Platform Specific Features:
Fixes #2943

### CODE REVIEW

**Comment by nezdolik** [comment]
cc @htuch we have chatted about this change long time ago prior to my parental leave.

**Comment by nezdolik** [comment]
cc @htuch we have chatted about this change long time ago prior to my parental leave.

**Comment by htuch** [comment]
Thanks so much @nezdolik, this is a definitely long asked for feature from the community and will be a welcome contribution. Assigning @adisuissa for review, since Adi is now leading work around xDS architecture in Envoy.

**Comment by adisuissa** [comment]
Apologies for the late reply.
Thanks for working on this, and thanks for trying this in prod and giving us the feedback!
Overall I think we would like to have this support for any xDS service, not only for EDS.

I'll start by giving some background on where I think we should be heading, and then answer your questions.

In the future, as we move towards the [new xDS-TP](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md#transport-network-addressing), the subscriptions to the same authority will essentially be multiplexed using this feature. Your optimized EDS stream allocation will work for each ADS subscription to a specific authority.
For now, we have a single gRPC stream when ADS is used (which doesn't need this multiplexing), and for non-ADS we have per-subscription stream.

IIUC the current design is: The cluster manager has a single `EdsSubscriptionFactory` which maps between config and a gRPC stream.
In the future, we will probably have a ConfigSubscriptionsManager in the server, that will own all the subscriptions, and a map between a config-source (and authority) to its actual GrpcMux. The GrpcMux can be configured to either SotW or Delta, and will abstract over it (just as is being done today).

For this PR I've got a few high-level comments to start with:
1. Should the EdsSubscriptionFactory extend SubscriptionFactoryImpl, and just have a method to return the correct GrpcMux (overriding the current create methods, and calling them if the subscription isn't there already)?
2. Can we use a different name for this factory, maybe `MultiplexedSubscriptionFactory`? (other ideas to make it more generic are welcome :) )
3. Can we add some integration tests?
4. For now I do agree that this factory should co-exist with the current (non-multiplexed) subscription factory, but we should gradually move away from using the original one. I think that introducing (and guarding) it for each xDS-type is the right way forward.

Addressing your questions:

> Do we need support for Delta Grpc APi type?
Yes, we should support both SotW and Delta, but I don't think it will add too much to this implementation.

> Do we need support for unified mux?
The unified-mux code essentially uses the same code-base for both SotW and Delta implementations, so yes we would like to support it. In the near future we will move to that code-base and abandon the old one. The unified mux code should implement the GrpcMux interface, so I don't think it will require much work.

> Implement collectionSubscriptionFromUrl method? (part of udpa/resource federation)
Yes, it should be implemented, but I think it is very straight-forward.

> Hide change behind feature flag in api? (as it is high risk)
I think this is highly desired, at least temporarily, as we want to allow the use of the current subscription code if we break anyone.

LMK if there's something unclear, and if I can help with anything.

**Comment by nezdolik** [comment]
@adisuissa thanks for a very detailed feedback and apologies for late reply, I have been stuck in figuring out mocks fix for the remaining Eds tests.
I agree that MultiplexedSubscriptionFactory should be a general feature that could be enabled for any XDS type (since RDS and SDS currently expose the same perf problem for a large number of resources).
We could start by implementing a generic MultiplexedSubscriptionFactory and exposing it as a config option in EDS api as a start, does that sound good?
I will start addressing your feedback in a series of commits and will ping you when all parts are done.
May need some help later with the integration test, I was not completely sure on which signal/stats to rely to verify that grpc channel is being multiplexed in the test(may figure it out later in the process).
The original idea was to look at Envoy stats for successfully completed requests for an H2 Eds cluster, since in H2 each stream maps to a single logical request/response.

**Comment by adisuissa** [comment]
> I agree that MultiplexedSubscriptionFactory should be a general feature that could be enabled for any XDS type (since RDS and SDS currently expose the same perf problem for a large number of resources).We could start by implementing a generic MultiplexedSubscriptionFactory and exposing it as a config option in EDS api as a start, does that sound good?
> 

Sounds good to me, thanks for all the efforts!
 
> May need some help later with the integration test, I was not completely sure on which signal/stats to rely to verify that grpc channel is being multiplexed in the test(may figure it out later in the process). The original idea was to look at Envoy stats for successfully completed requests for an H2 Eds cluster, since in H2 each stream maps to a single logical request/response.

I wonder if the [upstream statistics](https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats#general) can be used to count the number of connections.
If you point me to the specific test, I can try to take a look.

**Comment by nezdolik** [comment]
Applied portion of review comments:
* EdsSubscriptionFactory renamed to MultiplexedSubscriptionFactory
* MultiplexedSubscriptionFactory extends SubscriptionFactoryImpl and overrides method for creating mux
* MultiplexedSubscriptionFactory shares mux for subscriptions per management server per xds type url.
* Updated MultiplexedSubscriptionFactory test
I will now work on integration test, add tests for new feature and make it parametrised for xds type (sotw, delta) and grpc flavour (envoy, google).
@adisuissa I am not sure if upstream stats for number of connections is good enough signal to rely on. Since mux manages a single grpc stream (according to [docs](https://github.com/envoyproxy/envoy/blob/main/source/docs/xds.md)) and grpc channel may have many streams.

File: test/integration/eds_integration_test.cc
Comment (nezdolik) [inline test/integration/eds_integration_test.cc:897]: @adisuissa figured out the approach to verify the number of streams for both Grpc client types. For Envoy Grpc number of total streams towards specific cluster equals to total number of H2 requests. Google Grpc actually exposes stream metrics per client instance, which is nice. Please let me know wdyt.

File: test/integration/eds_integration_test.cc
Comment (nezdolik) [inline test/integration/eds_integration_test.cc:1]: to be cleaned up

**Comment by nezdolik** [comment]
* Added integration test to verify that discovery requests for multiple EDS clusters reuse one stream/mux towards same management server.

Looks like there is merge conflict that needs to be fixed. Will work now on adding api feature switch in EDS config and extending tests for the config property.

**Comment by nezdolik** [comment]
@adisuissa this is ready for review

**Comment by yanavlasov** [comment]
@adisuissa and @lizan this is ready for review

File: api/envoy/config/cluster/v3/cluster.proto
Comment (markdroth) [inline api/envoy/config/cluster/v3/cluster.proto:212]: I don't think this should be done by adding a new field to the API like this, especially not an EDS-specific field.  The fact that multiple EDS subscriptions are not currently multiplexed onto a single stream is basically a bug in Envoy, and I think we should just fix the bug.

If you explicitly want to find a way to make this configurable via the API, then a better approach would be to fix #13951.  That would allow a `ConfigSource` to specify whether it's an ADS stream, in which case Envoy will multiplex the EDS subscriptions onto a single ADS stream.  And that approach would work for all resource types, not just EDS.

Comment (nezdolik) [reply api/envoy/config/cluster/v3/cluster.proto:212]: @markdroth the config option was introduced to reduce risk of rolling out the change, as previously merged [fix](https://github.com/envoyproxy/envoy/pull/5272) has broken EDS for bunch of users. I agree is not the best usage of API, as eventually everyone should use the multiplexed version of subscription factory for EDS. Perhaps we could instead guard new feature with runtime flag? We have increased confidence for current bug fix, as it is now covered by integration tests and has been tested in prod environment, but there is still risk involved.

I would like to focus first on fixing EDS bug, as it has significant impact on us. I can further help out with https://github.com/envoyproxy/envoy/issues/13951 but as a separate effort.

Comment (markdroth) [reply api/envoy/config/cluster/v3/cluster.proto:212]: A runtime flag sounds fine.

I would really prefer not to add an EDS-specific option to the API for this, since it's really not an API issue, it's just an Envoy implementation bug.

**Comment by mattklein123** [comment]
Sorry just catching up on this issue. I would like to get this fixed for the folks at Spotify. Are we good with this overall change if we call it a bug and fix it under runtime flag? That seems fine to me. cc @markdroth @nezdolik

**Comment by markdroth** [comment]
Triggering this via a runtime flag instead of a new API field SGTM.

File: test/common/grpc/grpc_client_integration.h
Comment (nezdolik) [inline test/common/grpc/grpc_client_integration.h:176]: integration test for eds over grpc has cardinality of 16, it takes some time to run it.

**Comment by nezdolik** [comment]
@mattklein123 @markdroth switched to using runtime feature flag for eds multiplexing.

File: source/common/runtime/runtime_features.cc
Comment (mattklein123) [inline source/common/runtime/runtime_features.cc:97]: Is there any reason not to have this be true by default?

**Review by mattklein123 — COMMENTED** [review]
At a high level this makes sense to me but I will defer to @adisuissa for the full review. Thank you!

File: source/common/runtime/runtime_features.cc
Comment (nezdolik) [reply source/common/runtime/runtime_features.cc:97]: it makes sense, am just being overcautious

**Comment by phlax** [comment]
@adisuissa could you give this another pass please

File: source/common/runtime/runtime_features.cc
Comment (mattklein123) [reply source/common/runtime/runtime_features.cc:97]: If you can test it in prod ahead of time, I would flip to true by default. If not, I think it's fine to leave as false and come back and switch to true after you test it.

File: test/common/grpc/grpc_client_integration.h
Comment (adisuissa) [reply test/common/grpc/grpc_client_integration.h:176]: Can these tests be sharded?

File: test/common/grpc/grpc_client_integration.h
Comment (adisuissa) [inline test/common/grpc/grpc_client_integration.h:130]: Consider renaming to `MultiplexedDeltaSotwIntegrationParamTest` as this is not only for EDS.

File: test/integration/eds_integration_test.cc
Comment (adisuissa) [inline test/integration/eds_integration_test.cc:616]: Should be moved to `initializeTest()` as it not per-cluster behavior.

File: test/integration/eds_integration_test.cc
Comment (adisuissa) [inline test/integration/eds_integration_test.cc:921]: I'm wondering how difficult will it be to add a test that changes modes between "envoy.reloadable_features.multiplex_eds" enabled to disabled and back?

File: test/integration/eds_integration_test.cc
Comment (adisuissa) [inline test/integration/eds_integration_test.cc:918]: nit: just return at the beginning of the test if the edsUpdateMode isn't multiplexed.

File: test/integration/eds_integration_test.cc
Comment (adisuissa) [inline test/integration/eds_integration_test.cc:441]: If this class and tests do not share code with the above class/tests, I suggest refactoring it into a different file.

File: test/integration/eds_integration_test.cc
Comment (adisuissa) [inline test/integration/eds_integration_test.cc:592]: I may be missing something here, but shouldn't this test file verify that both the multiplexed and non-multiplexed are working correctly?
If so, shouldn't there be multiple xds streams, and sending updates on multiple streams (in the case of non-multiplexed), and use of a single stream in the multiplexed case?
Please let me know if I'm missing something.

**Review by adisuissa — COMMENTED** [review]
Apologies for the late reply, and thanks for working on this.
Overall the code LGTM, and added some comments on the tests.

I was considering the tradeoffs between your design, and having the multiplexed implementation as part of the current subscription factory implementation.
I think that it makes sense to use your current design, as it allows to gradually introduce the use for different xDS types. Eventually, I believe, we'll just use the multiplexed subscription for all gRPC-based xDS types.

File: test/integration/eds_integration_test.cc
Comment (nezdolik) [reply test/integration/eds_integration_test.cc:918]: this variable is needed for clean shutdown of the test, eg if test and its initialisation logic were skipped then cleanup logic (tear down xds connection, tear down conns to upstreams) is also skipped.

File: test/common/grpc/grpc_client_integration.h
Comment (nezdolik) [reply test/common/grpc/grpc_client_integration.h:176]: yes

File: test/integration/eds_integration_test.cc
Comment (nezdolik) [reply test/integration/eds_integration_test.cc:592]: That is correct. I intentionally omitted tests for multiple streams/clusters with no multiplexing, because the code is async and is not possible to predict the order of discovery requests. Is there is an established way to test out of order expected set of discovery requests/responses?

Comment (adisuissa) [reply test/integration/eds_integration_test.cc:592]: That's a great question.
I don't think there is something ready. I've implemented this once for delta-xDS in a previous PR (https://github.com/envoyproxy/envoy/pull/18195/commits/20b1674193e62b6087dd13340b7bc91c03de6301#diff-5227f6492ddd6badd76af87ce9f5f1c8acc349eed191e65a2bc05ab5d43b65f6R643) but eventually it wasn't needed for the tests.

File: test/common/grpc/grpc_client_integration.h
Comment (nezdolik) [reply test/common/grpc/grpc_client_integration.h:176]: after extracting eds grpc test to separate file, test running time have decreased, don't think we need to shard

File: test/integration/eds_integration_test.cc
Comment (nezdolik) [reply test/integration/eds_integration_test.cc]: @adisuissa it will require writing lots of code. Maybe not tricky, but large code wise

**Comment by nezdolik** [comment]
@adisuissa this is ready for review (applied review comments, added test for stream per cluster, enabled multiplexing by default). Only piece not done is integration test for turning on/off the flag via runtime.

File: test/extensions/clusters/eds/eds_test.cc
Comment (adisuissa) [inline test/extensions/clusters/eds/eds_test.cc:2609]:
```suggestion
TEST_F(EdsTest, MultiplexEdsEnabledByDefault) {
```

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Code LGTM, and I still need to go through the last integration test (sorry, it is quite large).
RE: "Only piece not done is integration test for turning on/off the flag via runtime", I thought that `MultiplexedDeltaSotwIntegrationParamTest` (which `EdsOverGrpcIntegrationTest` extends) takes care of that, no?

**Comment by nezdolik** [comment]
> Thanks for working on this! Code LGTM, and I still need to go through the last integration test (sorry, it is quite large). RE: "Only piece not done is integration test for turning on/off the flag via runtime", I thought that `MultiplexedDeltaSotwIntegrationParamTest` (which `EdsOverGrpcIntegrationTest` extends) takes care of that, no?

It does yes, what i meant is there is no test that turns flag on and then off via runtime in a single test.

**Comment by nezdolik** [comment]
This bug fix is important for our company as it unblocks significant cost cut down for cross-zone traffic for all services. 
We intended to use the fix early Q1, hopefully we can get it merged by then and not use internal fork with patch included.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by nezdolik** [comment]
The patch has been successfully tested in Spotify prod environment, @lizan

**Comment by KBaichoo** [comment]
friendly ping @lizan 

I think this is waiting for your input.

@nezdolik this needs a main merge

/wait

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

File: changelogs/current.yaml
Comment (mattklein123) [inline changelogs/current.yaml:127]: Can you say more about what this does? Also, can you flip the text about turning it off via runtime flag because I think this is on by default now?

This will need another rebase as we will land this after the release cut today due to risk. Thank you!

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
